go-controller/pkg/metrics/ovn_db.go

```diff
@@ -352,89 +349,85 @@ func getOvnDbVersionInfo() {
 	}
 }
 
-// ReadinessProbeCommand runs readiness probes against various targets
-var OvnDBExporterCommand = cli.Command{
-	Name:  "ovn-db-exporter",
-	Usage: "",
-	Flags: []cli.Flag{
-		&cli.StringFlag{
-			Name:  "metrics-bind-address",
-			Usage: `The IP address and port for the metrics server to serve on (default ":9476")`,
-		},
-	},
-	Action: func(ctx *cli.Context) error {
-		bindAddress := ctx.String("metrics-bind-address")
-		if bindAddress == "" {
-			bindAddress = "0.0.0.0:9476"
-		}
+func RegisterOvnDBMetrics(clientset *kubernetes.Clientset, k8sNodeName string) {
+	var podExists bool
 
-		if err := util.SetExec(kexec.New()); err != nil {
-			return err
+	err := wait.PollImmediate(1*time.Second, 300*time.Second, func() (bool, error) {
+		var err error
+		podExists, err = checkPodRunsOnGivenNode(clientset, "name=ovnkube-db", k8sNodeName)
+		if err != nil {
+			return false, nil
+		}
+		if podExists {
+			return true, nil
 		}
-		mux := http.NewServeMux()
-		mux.Handle("/metrics", promhttp.Handler())
-
-		// get the ovsdb server version info
-		getOvnDbVersionInfo()
-		// register metrics that will be served off of /metrics path
-		prometheus.MustRegister(metricOVNDBMonitor)
-		prometheus.MustRegister(metricOVNDBSessions)
-		prometheus.MustRegister(metricDBSize)
-		prometheus.MustRegister(prometheus.NewGaugeFunc(
-			prometheus.GaugeOpts{
-				Namespace: metrics.MetricOvnNamespace,
-				Subsystem: metrics.MetricOvnSubsystemDBRaft,
-				Name:      "build_info",
-				Help: "A metric with a constant '1' value labeled by ovsdb-server version and " +
-					"NB and SB schema version",
-				ConstLabels: prometheus.Labels{
-					"version":           ovnDbVersion,
-					"nb_schema_version": nbDbSchemaVersion,
-					"sb_schema_version": sbDbSchemaVersion,
-				},
+		return false, nil
+	})
+	if err != nil {
+		klog.Errorf("Timed out while checking if OVN DB Pod runs on this %q K8s Node: %v. "+
+			"Not registering OVN DB Metrics on this Node.", k8sNodeName, err)
+		return
+	}
+	if !podExists {
```

Comment:
can this happen? if pod does not exist err will be != nil and we return before this line

Reply:
@aojea the API will not return error if pod with set label doesn't exist. It will just return empty list. 

The only case the err will be non-nil is when it cannot communicate with the K8s API Server.

Reply:
`err := wait.PollImmediate(1*time.Second, 300*time.Second, func() (bool, error) {` 

will return an error `ErrWaitTimeout` if the function under evaluation does nor return true in 300 sec, i.e., if there are no pods and you want to differentiate you should check that the err returned is not `ErrWaitTimeout`

Reply:
@aojea in OVN K8s deployment there `must` be a pod whose label is ovnkube-db. You cannot have a cluster without OVN DB pods. In fact,  the onvkube-node Pod in which this code runs doesn't even start until the ovnkube-db is up and running.

So, there are two reasons why we will see error:
1. failure in the API call to List pods -- for this case we just log the error and re-try
2. we timeout after 300 seconds (the only reason for this time out is (1) above)

So, in both the cases above (1) and (2) we timeout and log that we timed out and couldn't determine if the Pod exists on the current node or not and move on.

Reply:
yeah, but my point is, if `podExists` is false then we always return `err != nil` ,
 so this ` if !podExists` will never match :smile: 
I'm I reading it wrong?

Reply:
@aojea You are right. I see what you are saying :). Can you please check the latest changes ?